### ISSUE TITLE
postgres: tolerate adding columns to upstream relations

### ISSUE THREAD

**Issue by petrosagg** [MaterializeInc/materialize#17595]
Postgres sources currently take a snapshot of the schema of the upstream postgres system and store it in the materialize catalog. During ingestion we monitor for events that signal schema changes and in those cases we use the snapshot we stored on creation to detect any deviations. If we find any for the tables we're replicating we immediately put the source in an errored state.

While that strategy is safe it is overly strict. This issue is about relaxing the schema changes checks of postgres sources such that we allow a user to `ALTER` an upstream table by adding a column. We can do so by doing a bit more work when we receive a notification that the schema changed.

Concretely, suppose that we are tasked to replicate some table with columns `[a, b, c]` and we receive a schema change notification and the table's schema is now `[a, c, d, b]`. Our source needs to be maintaining a projection vector that given a row from postgres produces the corresponding row of the schema in materialize. In this example the projection vector would start as just `[0, 1, 2]` but after the schema change we would alter it to `[0, 3, 1]`.

**Comment by sploiselle** [comment]
tl;dr [check out a WIP branch](https://github.com/sploiselle/materialize/tree/pg-source-add-drop-unused-cols)

As identified above, the main idea is to track the order to project the current
schema into the canonical schema and error iff that's not possible.

- Treat the schema we see when creating the source as canonical.
- In `postgres_replication_loop_inner`, check the publication tables (i.e. those
  currently in the publication) against the source tables (i.e. the canonical
  schemas), generating `BTreeMap<u32, Vec<usize>>`, which are the projections
  necessary to get the canonical schema out of the current schema for any given
  table. We'll refer to these as `projections`. We will generate projections
  only when necessary, i.e. tables that remain unchanged will not incur any
  overhead beyond checking for their ID in a `BTreeMap`.

  We'll perform this check unconditionally and outside the the `if` branch for
  producing a snapshot vs. reading from the replication stream. This will
  fast-forward us to handling the evolved schema if something changed while we
  were offline, as well as avoid us needing to persist the last schema we saw.

  Obviously, we'll error if the schema has changed in a way we don't support,
  e.g. dropping ingested columns, changing column names, etc.

- Provide `projections` when producing snapshots and data from the replicated
  stream. Both functions will learn to project data from tables with entries in
  `projections` into the proper order.

- `postgres_protocol::message::backend::LogicalReplicationMessage::Relation`
  messages will be compared against the source tables, performing the same check
  as we did originally to generate new `projection` entries. These entries will
  then be used as we receive subsequent data from the replication stream.

@petrosagg Open questions assuming the above design is viable:
- `storage::source::postgres::validate_tables` previously errored if columns'
  nullability or primary key membership changed; however, handling
  `LogicalReplicationMessage::Relation` messages doesn't give us that
  information.
  
  Should we:
  - Stop erroring on changes to those constraint
    - **Pro**: Makes handling schema changes consistent across reboot and in the replication stream
    - **Con**: Potentially starts generating junk data, e.g. if a column is no longer par
  - Try reconcile `LogicalReplicationMessage::Relation` with
  `mz_postgres_util::publication_info` to determine if a constraint changed
    - **Pro**: Prevents us from potentially footgunning the use of some invariant that changed
    - **Con**: Seems challenging in the face of possibly multiple schema changes;
      maybe you have some insight into how those are versioned in a way we can
      understand?
